Storage/Rust/RustRemoteTabs.swift

```diff
@@ -0,0 +1,171 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+import Shared
+@_exported import MozillaAppServices
+
+private let log = Logger.syncLogger
+
+public extension RemoteTabRecord {
+    func toRemoteTab(client: RemoteClient) -> RemoteTab? {
+        guard let url = Foundation.URL(string: self.urlHistory[0]) else {
+            return nil
+        }
+        let history = self.urlHistory[1...].map { url in Foundation.URL(string: url) }.compactMap { $0 }
+        let icon = self.icon != nil ? Foundation.URL(fileURLWithPath: self.icon ?? "") : nil
+
+        return RemoteTab(clientGUID: client.guid, URL: url, title: self.title, history: history, lastUsed: Timestamp(self.lastUsed), icon: icon)
+    }
+}
+
+public extension ClientRemoteTabs {
+    func toClientAndTabs(client: RemoteClient) -> ClientAndTabs {
+        return ClientAndTabs(client: client, tabs: self.remoteTabs.map { $0.toRemoteTab(client: client)}.compactMap { $0 })
+    }
+}
+
+public class RustRemoteTabs {
+    let databasePath: String
+
+    let queue: DispatchQueue
+    var storage: TabsStorage?
+
+    private(set) var isOpen = false
+
+    private var didAttemptToMoveToBackup = false
+
+    public init(databasePath: String) {
+        self.databasePath = databasePath
+
+        queue = DispatchQueue(label: "RustRemoteTabs queue: \(databasePath)", attributes: [])
+    }
+
+    private func open() -> NSError? {
+        storage = TabsStorage(databasePath: databasePath)
+        isOpen = true
+        return nil
+    }
+
+    private func close() -> NSError? {
+        storage = nil
+        isOpen = false
+        return nil
+    }
+
+    public func reopenIfClosed() -> NSError? {
+        var error: NSError?
+
+        queue.sync {
```

Comment:
I am seeing some queue.sync and at other places async and am trying to better understand its use case. 

Is it because `Queue.Sync` in reopenIfClosed gets called multiple times but we only want to run it in sequential order?

Reply:
I was following the pattern established with [RustLogins](https://github.com/mozilla-mobile/firefox-ios/blob/main/Storage/Rust/RustLogins.swift) and didn't think a whole lot about the use case. If this doesn't seem right to you, I'm open to changing it.

Reply:
Just adding a bit more context here. The native iOS `BrowserDB.reopenIfClosed` function, which is called with the new `reopenIfClosed` function for tabs in the [BrowserProfile reopen function](https://github.com/mozilla-mobile/firefox-ios/blob/197ef0e837380b24bca3070a892d1b52fa1784d7/Providers/Profile.swift#L371), also [submits DispatchQueue work items synchronously](https://github.com/mozilla-mobile/firefox-ios/blob/197ef0e837380b24bca3070a892d1b52fa1784d7/Storage/ThirdParty/SwiftData.swift#L240). So using `queue.sync` instead of `queue.async` is an established convention. The same is true for [BrowserDB.forceClose](https://github.com/mozilla-mobile/firefox-ios/blob/197ef0e837380b24bca3070a892d1b52fa1784d7/Storage/ThirdParty/SwiftData.swift#L223).

It also makes sense that the `reopenIfClosed` and `forceClose` function would block their respective thread pools since any subsequent work item submissions would fail. For instance if a `tabs.reopenIfClosed` was called but didn't complete before a subsequent call to a tabs CRUD method. In short I'm convinced `queue.sync` is the right call for these two functions.

Reply:
This is one area of PR which I am not sure about in terms of the benefits of the queue but I see your point of higher importance. 

For now, marking it as r+ and will think more about it and come back to it at a later time.